packages/opds-web-client/src/components/DownloadButton.tsx

```diff
@@ -1,132 +1,99 @@
+        window.open(url, "_blank");
+      });
     } else {
-      return this.props.fulfill(this.props.url).then(blob => {
-        download(
-          blob,
-          this.generateFilename(this.props.title),
-          // TODO: use mimeType variable once we fix the link type in our OPDS entries
-          this.mimeType()
-        );
+      // TODO: use mimeType variable once we fix the link type in our
+      // OPDS entries
+      action = actions.fulfillBook(url);
+      dispatchFn = dispatch(action).then(blob => {
+        download(blob, generateFilename(title), mimeTypeFn());
       });
     }
-  }
-
-  isIndirect() {
-    return (
-      this.props.indirectType &&
-      this.props.mimeType ===
-        "application/atom+xml;type=entry;profile=opds-catalog"
-    );
-  }
-
-  generateFilename(str: string): string {
-    return (
-      str
-        .toLowerCase()
-        .replace(/[^a-z0-9]+/g, "-")
-        .replace(/(^-|-$)/g, "") + this.fileExtension()
-    );
-  }
-
-  mimeType() {
-    return this.props.mimeType === "vnd.adobe/adept+xml"
+    return dispatchFn;
+  };
+  const isIndirect = () =>
+    indirectType &&
+    mimeType === "application/atom+xml;type=entry;profile=opds-catalog";
+  const generateFilename = (str: string): string =>
+    str
+      .toLowerCase()
+      .replace(/[^a-z0-9]+/g, "-")
+      .replace(/(^-|-$)/g, "") + fileExtension();
+  const mimeTypeFn = () =>
+    mimeType === "vnd.adobe/adept+xml"
       ? "application/vnd.adobe.adept+xml"
-      : this.props.mimeType;
-  }
-
-  fileExtension() {
-    return (
-      {
-        "application/epub+zip": ".epub",
-        "application/pdf": ".pdf",
-        "application/vnd.adobe.adept+xml": ".acsm",
-        "application/x-mobipocket-ebook": ".mobi"
-      }[this.mimeType()] || ""
-    );
-  }
-
-  downloadLabel() {
+      : mimeType;
+  const fileExtension = () =>
+    ({
+      "application/epub+zip": ".epub",
+      "application/pdf": ".pdf",
+      "application/vnd.adobe.adept+xml": ".acsm",
+      "application/x-mobipocket-ebook": ".mobi"
+    }[mimeTypeFn()] || "");
+  const downloadLabel = () => {
     if (
-      this.props.indirectType ===
+      indirectType ===
       "text/html;profile=http://librarysimplified.org/terms/profiles/streaming-media"
     ) {
       return "Read Online";
     }
-    let type = this.fileExtension()
+    let type = fileExtension()
       .replace(".", "")
       .toUpperCase();
-    return "Download" + (type ? " " + type : "");
+    return `Download${type ? " " + type : ""}`;
+  };
+  let downloadProps = {
+    className: "btn btn-default download-button",
+    ...elementProps
+  };
+  if (isPlainLink) {
+    downloadProps["href"] = url;
+    downloadProps["target"] = "_blank";
+  } else {
+    downloadProps["onClick"] = fulfill;
+    downloadProps["className"] =
+      `${downloadProps["className"]} ` +
+      `download-${fileExtension().slice(1)}-button`;
   }
-}
+
+  return (
+    <span>
+      {React.createElement(
```

Comment:
I would replace this with just a ternary render, and then you can also get rid of the if/else statement above. Something like this:
```js
<span>
    {isPlainLink ? 
        <a className="..." href={url} target="_blank" /** your other props */ />  : 
        <button /** props for the button version */ /> 
    }
</span>
```

Reply:
That seems simpler to read and reason about what props are going where